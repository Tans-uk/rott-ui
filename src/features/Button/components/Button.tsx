import {type FC} from 'react'

import {ActivityIndicator, StyleSheet, TouchableOpacity} from 'react-native'

import type {ButtonProps} from '../models'
import {ButtonStyles} from '../styles'
import {buttonSizeNormalizer} from '../utils'

import {Icon, type IconTypes} from '@features/Icon'
import {Image} from '@features/Image'
import {Label} from '@features/Label'
import {type Size} from '@models'
import {themeConfig} from '@providers'
import {display} from '@utils'

export const Button: FC<ButtonProps> = ({
  variant = 'primary',
  color,
  size = {height: 'lg'},
  fontWeight,
  fontFamily,
  isLoading,
  loadingText,
  flex = (typeof size === 'object' && size?.width) || typeof size === 'string' ? 0 : 1,

  borderWidth,
  borderRadius,
  borderColor,
  flexGrow,
  circle = false,

  leftImage,
  rightImage,

  leftIcon,
  rightIcon,

  height = 'full',
  width = 'full',
  justifyContentFlexEnd,
  justifyContentFlexStart,
  justifyContentSpaceAround,
  justifyContentSpaceBetween,
  justifyContentCenter = justifyContentFlexEnd ||
  justifyContentFlexStart ||
  justifyContentSpaceAround ||
  justifyContentSpaceBetween
    ? false
    : true,

  children,
  style,
  text,
  ...props
}) => {
  return (
    <TouchableOpacity
      activeOpacity={0.8}
      style={StyleSheet.flatten([
        ButtonStyles({
          variant,
          color,
          size,
          flex,
          flexGrow,
          borderWidth,
          borderRadius,
          borderColor,
          circle,
          width,
          height,
          includeJustifyContent: true,
          justifyContentCenter,
          justifyContentFlexEnd,
          justifyContentFlexStart,
          justifyContentSpaceAround,
          justifyContentSpaceBetween,
          ...props,
        }).defaultButtonStyle,
        style,
      ])}
      {...props}>
      {leftIcon && !isLoading && (
        <Icon
          {...leftIcon}
          absolute={leftIcon.absolute}
          left={leftIcon.absolute ? display.normalize(24) : undefined}
          testID='button-left-icon-test-id'
          marginRight={leftIcon.absolute ? 0 : 8}
          variant={leftIcon?.variant}
          name={leftIcon?.name as IconTypes}
          width={
            leftIcon?.width ??
            buttonSizeNormalizer(typeof size === 'string' ? size : (size.height as Size)).icon
          }
          height={
            leftIcon?.height ??
            buttonSizeNormalizer(typeof size === 'string' ? size : (size.height as Size)).icon
          }
        />
      )}

      {leftImage && !isLoading && (
        <Image
          testID='button-left-image-test-id'
          left={leftImage.absolute ? display.normalize(24) : undefined}
          name={leftImage.name}
          absolute={leftImage.absolute}
          marginRight={leftImage.absolute ? 0 : 8}
          width={
            leftImage.width ??
            buttonSizeNormalizer(typeof size === 'string' ? size : (size.height as Size)).icon
          }
          height={
            leftImage.height ??
            buttonSizeNormalizer(typeof size === 'string' ? size : (size.height as Size)).icon
          }
          tintColor={leftImage.tintColor}
          resizeMode={leftImage.resizeMode}
        />
      )}

      {!isLoading && (
        <Label
          numberOfLines={1}
          textCenter
          fontFamily={(fontFamily ?? !fontWeight) ? 'Markpro-Bold' : undefined}
          fontSize={
            typeof size === 'object'
              ? buttonSizeNormalizer(size.height as Size).fontSize
              : buttonSizeNormalizer(size).fontSize
          }
          fontWeight={fontWeight}
          style={ButtonStyles({color, variant}).buttonTextStyle}>
          {children ?? text}
        </Label>
      )}

      {isLoading && (
        <ActivityIndicator
          color={themeConfig.colors.white}
          testID='button-loading-indicator-test-id'
        />
      )}

      {isLoading && loadingText && (
        <Label
          testID='button-loading-text-test-id'
          style={ButtonStyles({color, variant}).buttonTextStyle}
          marginLeft={8}>
          {loadingText}
        </Label>
      )}

      {rightIcon && !isLoading && (
        <Icon
          {...rightIcon}
          absolute={rightIcon.absolute}
          right={rightIcon.absolute ? display.normalize(24) : undefined}
          testID='button-right-icon-test-id'
          marginLeft={rightIcon.absolute ? 0 : 8}
          variant={rightIcon?.variant}
          name={rightIcon?.name as IconTypes}
          width={
            rightIcon.width ??
            buttonSizeNormalizer(typeof size === 'string' ? size : (size.height as Size)).icon
          }
          height={
            rightIcon.height ??
            buttonSizeNormalizer(typeof size === 'string' ? size : (size.height as Size)).icon
          }
        />
      )}

      {rightImage && !isLoading && (
        <Image
          testID='button-right-image-test-id'
          right={rightImage.absolute ? display.normalize(24) : undefined}
          name={rightImage.name}
          absolute={rightImage.absolute}
          marginRight={rightImage.absolute ? 0 : 8}
          width={
            rightImage.width ??
            buttonSizeNormalizer(typeof size === 'string' ? size : (size.height as Size)).icon
          }
          height={
            rightImage.height ??
            buttonSizeNormalizer(typeof size === 'string' ? size : (size.height as Size)).icon
          }
          tintColor={rightImage.tintColor}
          resizeMode={rightImage.resizeMode}
        />
      )}
    </TouchableOpacity>
  )
}
