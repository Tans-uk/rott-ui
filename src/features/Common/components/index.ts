export * from './CommonItem'
export * from './CommonItemContainer'
