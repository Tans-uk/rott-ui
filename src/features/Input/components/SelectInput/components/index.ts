export * from './SelectInput'
export * from './SelectInputModalComponent'
