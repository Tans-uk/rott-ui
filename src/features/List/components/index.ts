export * from './List'
export * from './FavoriteSwipeComponent'
export * from './ListSkeletonItem'
