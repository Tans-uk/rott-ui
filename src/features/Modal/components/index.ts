export * from './ModalComponent'
export * from './ModalContainer'
