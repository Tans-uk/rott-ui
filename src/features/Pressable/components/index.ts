export * from './Pressable'
export * from './PressableContainer'
