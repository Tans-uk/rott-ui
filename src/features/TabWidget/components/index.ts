export * from './TabWidget'
export * from './TabWidgetItem'
