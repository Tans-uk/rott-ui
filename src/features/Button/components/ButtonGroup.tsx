import {type FC} from 'react'

import {Platform} from 'react-native'

import {useSafeArea} from '../../../hooks'
import {themeConfig} from '../../../providers'
import {Content, type ContentProps} from '../../Content'
import {Image, type ImageProps} from '../../Image'
import {Item} from '../../Item'
import type {ButtonProps} from '../models'
import {Button} from './Button'
import React from 'react'

interface ButtonGroupProps extends ContentProps {
  buttons: ButtonProps | Nullable<ButtonProps>[]
  hideButtons?: boolean
  sticky?: boolean
  image?: ImageProps
  children?: any
  isFastTransfer?: boolean
}

/**
 *
 * @param hideButtons - Butonları gizler, !!! TODO: Daha sonra daha iyi bir çözüm bulunabilir mi bakılmalı!
 * @returns
 */
export const ButtonGroup: FC<ButtonGroupProps> = ({
  buttons,
  hideButtons,
  sticky,
  image,
  row,
  gap = 16,
  isFastTransfer,
  ...props
}) => {
  /**
   * Ekranın altındaki NavBar ile çakışan Content'lerde (Örn:ButtonGroup) otomatik olarak 24 birim padding ekler.
   * IOS'te navbar transparan olduğu için eğer Navbar var ise 24 birim eklenmez. Çünkü kendisi boşluk gibi görünür.
   *
   * Etkilenen Ekranlar: LoginScreen, TransferToMobileScreen
   */
  const {bottom} = useSafeArea()
  let paddingBottom = bottom

  if (Platform.OS === 'android' || bottom === 0) paddingBottom += 24

  const hasMoreThanOneButton = Array.isArray(buttons)
  const stickyBottomProps = {
    shadowColor: themeConfig.colors['neutral-shadow-300'],
    shadowOffset: {width: 0, height: -8},
    shadowOpacity: 1,
    shadowRadius: 35,
    borderTopWidth: 1,
    borderTopColor: themeConfig.colors['neutral-grey-alpha-200'],
  }

  return (
    <Content
      flexShrink={1}
      paddingTop={24}
      paddingBottom={paddingBottom}
      useBottomInset
      {...props}
      {...(sticky ? stickyBottomProps : {})}>
      {!hideButtons && (
        <Item row={row} gap={gap}>
          {hasMoreThanOneButton &&
            buttons
              .filter((button) => button)
              .map((button, index) => (
                <Button key={button?.testID ?? `${button?.text}_${index}`} {...button} />
              ))}
          {!hasMoreThanOneButton && <Button {...buttons} />}
        </Item>
      )}

      {(!!image || isFastTransfer) && (
        <Item justifyContentCenter alignItemsCenter marginTop={16}>
          <Image
            testID='image-test-id'
            {...image}
            width={isFastTransfer ? 45 : image?.width}
            height={isFastTransfer ? 30 : image?.height}
            name={isFastTransfer ? 'FAST_LOGO' : image?.name}
          />
        </Item>
      )}
    </Content>
  )
}
