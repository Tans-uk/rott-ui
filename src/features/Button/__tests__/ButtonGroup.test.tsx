import {ButtonGroup} from '../components'

import {render} from '@utils'

describe('ButtonGroup -> Custom Component', () => {
  const testIDs = {
    logoImageTestId: 'image-test-id',
  }

  it('buton grubunun snapshotı ile eşleşmeli', () => {
    const rendered = render(
      <ButtonGroup
        sticky
        buttons={[
          {text: 'Button 1', variant: 'primary'},
          {text: 'Button 2', variant: 'secondary-outline'},
        ]}
        image={{
          name: 'FAST_LOGO',
          width: 60,
          height: 40,
        }}
      />
    )

    expect(rendered).toMatchSnapshot()
  })

  it('isFastTransfer true ise ekranda fast görseli görünür', () => {
    const {logoImageTestId} = testIDs
    const {getByTestId} = render(
      <ButtonGroup
        isFastTransfer
        buttons={[
          {text: 'Button 1', variant: 'primary'},
          {text: 'Button 2', variant: 'secondary-outline'},
        ]}
      />
    )

    const logoImage = getByTestId(logoImageTestId)

    expect(logoImage).toBeOnTheScreen()
  })
})
