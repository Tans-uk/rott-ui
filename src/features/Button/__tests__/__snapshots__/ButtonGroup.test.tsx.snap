// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ButtonGroup -> Custom Component buton grubunun snapshotı ile eşleşmeli 1`] = `
<View
  flex={1}
  pointerEvents="box-none"
  style={
    {
      "flex": 1,
      "flexDirection": "column",
    }
  }
>
  <View
    borderTopColor="rgba(234, 239, 240, 0.15)"
    borderTopWidth={1}
    flexShrink={1}
    paddingBottom={24}
    paddingTop={24}
    shadowColor="rgba(6, 14, 16, 0.30)"
    shadowOffset={
      {
        "height": -8,
        "width": 0,
      }
    }
    shadowOpacity={1}
    shadowRadius={35}
    style={
      {
        "flex": undefined,
        "flexDirection": "column",
        "flexShrink": 1,
        "paddingBottom": 46,
        "paddingHorizontal": 46,
        "paddingTop": 46,
      }
    }
  >
    <View
      style={
        {
          "flex": undefined,
          "flexDirection": "column",
          "gap": 16,
        }
      }
    >
      <View
        accessibilityState={
          {
            "busy": undefined,
            "checked": undefined,
            "disabled": undefined,
            "expanded": undefined,
            "selected": undefined,
          }
        }
        accessibilityValue={
          {
            "max": undefined,
            "min": undefined,
            "now": undefined,
            "text": undefined,
          }
        }
        accessible={true}
        collapsable={false}
        focusable={false}
        onClick={[Function]}
        onResponderGrant={[Function]}
        onResponderMove={[Function]}
        onResponderRelease={[Function]}
        onResponderTerminate={[Function]}
        onResponderTerminationRequest={[Function]}
        onStartShouldSetResponder={[Function]}
        style={
          {
            "alignItems": "center",
            "backgroundColor": "rgba(0, 169, 206, 1)",
            "borderColor": "white",
            "borderRadius": 16,
            "borderWidth": undefined,
            "color": undefined,
            "flex": 1,
            "flexDirection": "row",
            "height": 108,
            "justifyContent": "center",
            "opacity": 1,
            "position": "relative",
            "width": 658,
          }
        }
      >
        <Text
          numberOfLines={1}
          style={
            {
              "color": "rgba(255, 255, 255, 1)",
              "flexShrink": 1,
              "fontFamily": "Markpro-Bold",
              "fontSize": 26,
              "fontWeight": undefined,
              "textAlign": "center",
              "textAlignVertical": "center",
            }
          }
        >
          Button 1
        </Text>
      </View>
      <View
        accessibilityState={
          {
            "busy": undefined,
            "checked": undefined,
            "disabled": undefined,
            "expanded": undefined,
            "selected": undefined,
          }
        }
        accessibilityValue={
          {
            "max": undefined,
            "min": undefined,
            "now": undefined,
            "text": undefined,
          }
        }
        accessible={true}
        collapsable={false}
        focusable={false}
        onClick={[Function]}
        onResponderGrant={[Function]}
        onResponderMove={[Function]}
        onResponderRelease={[Function]}
        onResponderTerminate={[Function]}
        onResponderTerminationRequest={[Function]}
        onStartShouldSetResponder={[Function]}
        style={
          {
            "alignItems": "center",
            "backgroundColor": "transparent",
            "borderColor": "rgba(255, 199, 44, 1)",
            "borderRadius": 16,
            "borderWidth": 2,
            "color": undefined,
            "flex": 1,
            "flexDirection": "row",
            "height": 108,
            "justifyContent": "center",
            "opacity": 1,
            "position": "relative",
            "width": 658,
          }
        }
      >
        <Text
          numberOfLines={1}
          style={
            {
              "color": "rgba(255, 199, 44, 1)",
              "flexShrink": 1,
              "fontFamily": "Markpro-Bold",
              "fontSize": 26,
              "fontWeight": undefined,
              "textAlign": "center",
              "textAlignVertical": "center",
            }
          }
        >
          Button 2
        </Text>
      </View>
    </View>
    <View
      alignItemsCenter={true}
      justifyContentCenter={true}
      marginTop={16}
      style={
        {
          "alignItems": "center",
          "flex": undefined,
          "flexDirection": "column",
          "justifyContent": "center",
          "marginTop": 31,
        }
      }
    >
      <Image
        height={40}
        source={
          {
            "testUri": "../../../src/assets/images/fast/fast-tcmb.png",
          }
        }
        style={
          {
            "height": 77,
            "maxHeight": 77,
            "overflow": "hidden",
            "resizeMode": "contain",
            "width": 116,
          }
        }
        testID="image-test-id"
        width={60}
      />
    </View>
  </View>
  <View
    onLayout={[Function]}
    pointerEvents="box-none"
    style={
      [
        {
          "alignSelf": "center",
          "elevation": 999999,
          "flex": 0,
          "maxWidth": "100%",
          "position": "absolute",
          "zIndex": 999999,
        },
        {
          "flexDirection": "column-reverse",
          "justifyContent": "flex-start",
          "top": 30,
          "width": 750,
        },
      ]
    }
  >
    <View
      style={
        {
          "bottom": 0,
        }
      }
    >
      <RCTSafeAreaView />
    </View>
  </View>
  <View
    onLayout={[Function]}
    pointerEvents="box-none"
    style={
      [
        {
          "alignSelf": "center",
          "elevation": 999999,
          "flex": 0,
          "maxWidth": "100%",
          "position": "absolute",
          "zIndex": 999999,
        },
        {
          "bottom": 40,
          "flexDirection": "column",
          "justifyContent": "flex-end",
          "width": 750,
        },
      ]
    }
  >
    <View
      style={
        {
          "bottom": 0,
        }
      }
    >
      <RCTSafeAreaView />
    </View>
  </View>
</View>
`;
